hack/helm/release/internal/helm/helm.go

```diff
@@ -0,0 +1,616 @@
+	for _, n := range names {
+		if n == name {
+			return true
+		}
+	}
+	return false
+}
+
+// uploadChartsConfig is the configuration for uploading Helm charts
+// and updating the GCS bucket Helm index file.
+type uploadChartsConfig struct {
+	releaseConf ReleaseConfig
+	noDeps      charts
+	withDeps    charts
+}
+
+// uploadChartsAndUpdateIndex will perform the following:
+// 1. Upload Helm charts with no dependencies to GCS bucket.
+// 2. Potentially update GCS Bucket Helm index.
+// 3. Potentially upload Helm charts with dependencies to GCS bucket.
+// 4. Potentially update GCS Bucket Helm index a second time.
+func uploadChartsAndUpdateIndex(conf uploadChartsConfig) error {
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer cancel()
+	tempDir, err := os.MkdirTemp(os.TempDir(), "charts")
+	if err != nil {
+		return fmt.Errorf("while creating temporary directory for charts: %w", err)
+	}
+	defer os.RemoveAll(tempDir)
+	// upload charts without dependencies
+	if err := uploadCharts(ctx, tempDir, conf.noDeps, conf.releaseConf); err != nil {
+		return err
+	}
+	if err := updateIndex(ctx, tempDir, conf); err != nil {
+		return err
+	}
+	// This retry is here because of caching in front of the Helm repository
+	// and the time it takes for a new release to show up in the repository.
+	// If the eck-stack chart depends on new version of any of the other
+	// eck-resources charts, and that new version is just released, then
+	// it will take ~ one hour for it to show up, so we will continue trying
+	// to get all dependencies of the helm charts, and upload them for 1 hour.
+	retry.Do(
+		func() error {
+			if err := updateHelmRepositories(); err != nil {
+				return err
+			}
+			// upload charts with dependencies
+			if err := uploadCharts(ctx, tempDir, conf.withDeps, conf.releaseConf); err != nil {
+				return err
+			}
+			if err := updateIndex(ctx, tempDir, conf); err != nil {
+				return err
+			}
+
+			return nil
+		},
+		retry.RetryIf(func(err error) bool {
+			if strings.Contains(err.Error(), "while updating dependencies for helm chart") && !conf.releaseConf.DryRun {
+				return true
+			}
+			return false
+		}),
+		retry.Attempts(60),
+		retry.Delay(1*time.Minute),
+		retry.OnRetry(func(n uint, err error) {
+			log.Printf("retry #%d: %s\n", n, err)
+		}),
+	)
+
+	return nil
+}
+
+func updateHelmRepositories() error {
+	// simulates helm repo update
+	f, err := repo.LoadFile(cli.New().RepositoryConfig)
+	if err != nil {
+		return fmt.Errorf("while loading repository file: %w", err)
+	}
+	for _, r := range f.Repositories {
+		cr, err := repo.NewChartRepository(r, getter.All(cli.New()))
+		if err != nil {
+			return fmt.Errorf("while getting new chart repository: %w", err)
+		}
+		if _, err = cr.DownloadIndexFile(); err != nil {
+			return fmt.Errorf("while updating repository index for (%s): %w", cr.Config.Name, err)
+		}
+		log.Printf("Updated repository index for (%s)\n", cr.Config.Name)
```

Comment:
I've tested to release in `--env=dev` and I see that we are updating the 2 repositories, is it normal?:

```
2023/01/27 20:32:08 Updated repository index for (elastic) elastic
2023/01/27 20:32:08 Updated repository index for (elastic-dev) elastic-dev
```

Reply:
I'm assuming you ran this locally?  It's loading your local configuration, which includes both elastic and elastic-dev repositories, and updating all the indexes, which is a harmless operation.  You won't see this in Buildkite.

```
func updateHelmRepositories() error {
	// simulates helm repo update
	f, err := repo.LoadFile(cli.New().RepositoryConfig)
	if err != nil {
		return fmt.Errorf("while loading repository file: %w", err)
	}
	for _, r := range f.Repositories {
		cr, err := repo.NewChartRepository(r, getter.All(cli.New()))
		if err != nil {
			return fmt.Errorf("while getting new chart repository: %w", err)
		}
		if _, err = cr.DownloadIndexFile(); err != nil {
			return fmt.Errorf("while updating repository index for (%s): %w", cr.Config.Name, err)
		}
		log.Printf("Updated repository index for (%s)\n", cr.Config.Name)
	}
	return nil
}
```

For me, it's even a bit more:
```
2023/01/30 13:58:20 Updated repository index for (grafana)
2023/01/30 13:58:21 Updated repository index for (metrics-server)
2023/01/30 13:58:22 Updated repository index for (prometheus-community)
2023/01/30 13:58:23 Updated repository index for (elastic)
2023/01/30 13:58:24 Updated repository index for (jetstack)
2023/01/30 13:58:24 Updated repository index for (cdf)
2023/01/30 13:58:25 Updated repository index for (argo)
2023/01/30 13:58:25 Updated repository index for (nginx-stable)
2023/01/30 13:58:25 Updated repository index for (aws-ebs-csi-driver)
2023/01/30 13:58:26 Updated repository index for (elastic-helm-dev)
```